Possible optimization of RingMined event

Brechtpd opened the pull request:
This saves about ~4000 gas. 

You may not like everything I did to do this. I minimized the data in the event like this:
- the data is sent in a way that still allows reconstructing the values, though some logic is needed
- lrcReward and lrcFee is packed in a single int called lrcBalance. lrcBalance will be positive when lrcReward is non-zero and negative when lrcFee is non-zero (lrcReward and lrcFee cannot be both non-zero).
- no separate array for the order hashes

lrcBalance is also used to send less data to batchTransferToken for some small extra gas savings.

lrcReward and lrcFee could also be stored as lrcBalance directly in the OrderState struct for some memory savings, though I'm not sure if that's a good idea for readability

Let me know if you're okay with any of these changes. I'm aware not all tests succeed because the interface of batchTransferToken has changed. If necessary I will of course update those tests.

dong77 commented:
@Brechtpd There were some major changes to the master branch that require this PR to resolve conflicts, again. Sorry about that. We had to merge that PR as it has some change in the overall design of the protocol.

Brechtpd commented:
Updated branch. Gas savings still ~4000 (452.500 -> 448.500).
I removed the change to batchTransferToken because I feel like the ~500 gas saving wasn't big enough.

kongliangzhong commented:
After fixing the conflicts, the unit test breaks. Sorry again, may you fix this again, thanks!

Brechtpd commented:
Updated the PR. I removed the lrcReward/lrcFee in lrcBalance thing because the code didn't look nice enough for the ~500 gas savings it did. Gas savings should now we a bit more than ~3000 (470500 -> 467500).

dong77 commented:
### PREVIOUSLY:
- cumulativeGasUsed for a ring of 2 orders: 392449
- cumulativeGasUsed for a ring of 3 orders: 470644

### NOW:
